shm_ib/test_ucp_rma_mt.put_get/1 failure on legacy test node

```
2020-04-08T16:12:07.6348612Z [----------] 1 test from shm_ib/test_ucp_rma_mt
2020-04-08T16:12:07.6349779Z [ RUN      ] shm_ib/test_ucp_rma_mt.put_get/1 <shm,ib>
2020-04-08T16:12:24.1050625Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1158005Z Value of: target_data[i]
2020-04-08T16:12:24.1159851Z   Actual: 16045690984833335043
2020-04-08T16:12:24.1160627Z Expected: orig_data[i]
2020-04-08T16:12:24.1161313Z Which is: 0
2020-04-08T16:12:24.1162965Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1163966Z Value of: target_data[i]
2020-04-08T16:12:24.1164695Z   Actual: 16045690984833335023
2020-04-08T16:12:24.1165436Z Expected: orig_data[i]
2020-04-08T16:12:24.1166116Z Which is: 0
2020-04-08T16:12:24.1167543Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1168549Z Value of: target_data[i]
2020-04-08T16:12:24.1169276Z   Actual: 16045690984833335033
2020-04-08T16:12:24.1170017Z Expected: orig_data[i]
2020-04-08T16:12:25.7996278Z Which is: 0
2020-04-08T16:12:25.7997756Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:25.7998675Z Value of: target_data[i]
2020-04-08T16:12:25.7999078Z   Actual: 16045690984833335053
2020-04-08T16:12:25.7999461Z Expected: orig_data[i]
2020-04-08T16:12:25.7999801Z Which is: 0
2020-04-08T16:12:25.8000658Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:197: Failure
2020-04-08T16:12:25.8001276Z Error: Endpoint timeout
2020-04-08T16:12:25.8002134Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:197: Failure
2020-04-08T16:12:25.8002726Z Error: Endpoint timeout
2020-04-08T16:12:25.8003516Z terminate called after throwing an instance of 'ucs::test_abort_exception'
2020-04-08T16:12:25.8004075Z   what():  std::exception
2020-04-08T16:12:25.8004944Z [swx-rdmz-ucx-legacy-01:57501:1:3010] Caught signal 6 (Aborted: tkill(2) or tgkill(2))
2020-04-08T16:12:25.8005567Z terminate called recursively
2020-04-08T16:12:25.8005938Z make: *** [test] Aborted
2020-04-08T16:12:25.8006825Z make: Leaving directory `/scrap/azure/agent-09/AZP_WORKSPACE/2/s/build-test/test/gtest'
```
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=5326&view=logs&j=88d29b50-4e11-53d5-399b-e634502852e6&t=cb773578-1ba5-52b8-cb10-437b22e73aab
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/5326/logs/95

--- a/src/uct/ib/dc/dc_mlx5.c
+++ b/src/uct/ib/dc/dc_mlx5.c
@@ -304,7 +304,7 @@ static ucs_status_t uct_dc_mlx5_iface_create_qp(uct_dc_mlx5_iface_t *iface,
                                                 uct_dc_dci_t *dci)
 {
     uct_ib_iface_t *ib_iface           = &iface->super.super.super;
-    uct_ib_qp_attr_t attr              = {};
+    uct_ib_mlx5_qp_attr_t attr         = {};
     ucs_status_t status;
 #if HAVE_DC_DV
     uct_ib_device_t *dev               = uct_ib_iface_device(ib_iface);
@@ -319,13 +319,13 @@ static ucs_status_t uct_dc_mlx5_iface_create_qp(uct_dc_mlx5_iface_t *iface,
         return status;
     }
 
-    uct_ib_iface_fill_attr(ib_iface, &attr);
-    attr.ibv.cap.max_recv_sge           = 0;
+    uct_ib_iface_fill_attr(ib_iface, &attr.super);
+    attr.super.ibv.cap.max_recv_sge     = 0;
 
     dv_attr.comp_mask                   = MLX5DV_QP_INIT_ATTR_MASK_DC;
     dv_attr.dc_init_attr.dc_type        = MLX5DV_DCTYPE_DCI;
     dv_attr.dc_init_attr.dct_access_key = UCT_IB_KEY;
-    qp = mlx5dv_create_qp(dev->ibv_context, &attr.ibv, &dv_attr);
+    qp = mlx5dv_create_qp(dev->ibv_context, &attr.super.ibv, &dv_attr);
     if (qp == NULL) {
         ucs_error("mlx5dv_create_qp("UCT_IB_IFACE_FMT", DCI): failed: %m",
                   UCT_IB_IFACE_ARG(ib_iface));
@@ -374,7 +374,7 @@ static ucs_status_t uct_dc_mlx5_iface_create_qp(uct_dc_mlx5_iface_t *iface,
     }
 
     uct_rc_txqp_available_set(&dci->txqp, dci->txwq.bb_max);
-    *cap = attr.ibv.cap;
+    *cap = attr.super.ibv.cap;
     return UCS_OK;
 
 err:

--- a/src/uct/ib/mlx5/dv/ib_mlx5_dv.c
+++ b/src/uct/ib/mlx5/dv/ib_mlx5_dv.c
@@ -37,7 +37,7 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
 ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
                                         uct_ib_mlx5_qp_t *qp,
                                         uct_ib_mlx5_txwq_t *tx,
-                                        uct_ib_qp_attr_t *attr)
+                                        uct_ib_mlx5_qp_attr_t *attr)
 {
     uct_ib_mlx5_md_t *md   = ucs_derived_of(iface->super.md, uct_ib_mlx5_md_t);
     uct_ib_device_t *dev   = &md->super.dev;
@@ -57,7 +57,7 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
     void *qpc;
     int ret;
 
-    uct_ib_iface_fill_attr(iface, attr);
+    uct_ib_iface_fill_attr(iface, &attr->super);
 
     uar = uct_worker_tl_data_get(iface->super.worker,
                                  UCT_IB_MLX5_DEVX_UAR_KEY,
@@ -74,12 +74,12 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
                sizeof(struct mlx5_wqe_umr_ctrl_seg) +
                sizeof(struct mlx5_wqe_mkey_context_seg) +
                ucs_max(sizeof(struct mlx5_wqe_umr_klm_seg), 64) +
-               ucs_max(attr->cap.max_send_sge * sizeof(struct mlx5_wqe_data_seg),
+               ucs_max(attr->super.cap.max_send_sge * sizeof(struct mlx5_wqe_data_seg),
                        ucs_align_up(sizeof(struct mlx5_wqe_inl_data_seg) +
-                                    attr->cap.max_inline_data, 16));
-    len_tx = ucs_roundup_pow2_or0(attr->cap.max_send_wr * wqe_size);
+                                    attr->super.cap.max_inline_data, 16));
+    len_tx = ucs_roundup_pow2_or0(attr->super.cap.max_send_wr * wqe_size);
     max_tx = len_tx / MLX5_SEND_WQE_BB;
-    max_rx = ucs_roundup_pow2_or0(attr->cap.max_recv_wr);
+    max_rx = ucs_roundup_pow2_or0(attr->super.cap.max_recv_wr);
     len    = len_tx + max_rx * UCT_IB_MLX5_MAX_BB * UCT_IB_MLX5_WQE_SEG_SIZE;
 
     if (tx != NULL) {
@@ -104,13 +104,13 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
         goto err_free_mem;
     }
 
-    dv.pd.in  = attr->ibv.pd;
+    dv.pd.in  = attr->super.ibv.pd;
     dv.pd.out = &dvpd;
-    dv.cq.in  = attr->ibv.send_cq;
+    dv.cq.in  = attr->super.ibv.send_cq;
     dv.cq.out = &dvscq;
     dvflags   = MLX5DV_OBJ_PD | MLX5DV_OBJ_CQ;
     mlx5dv_init_obj(&dv, dvflags);
-    dv.cq.in  = attr->ibv.recv_cq;
+    dv.cq.in  = attr->super.ibv.recv_cq;
     dv.cq.out = &dvrcq;
     mlx5dv_init_obj(&dv, MLX5DV_OBJ_CQ);
 
@@ -120,16 +120,16 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
     UCT_IB_MLX5DV_SET(qpc, qpc, pm_state, UCT_IB_MLX5_QPC_PM_STATE_MIGRATED);
     UCT_IB_MLX5DV_SET(qpc, qpc, pd, dvpd.pdn);
     UCT_IB_MLX5DV_SET(qpc, qpc, uar_page, uar->uar->page_id);
-    ucs_assert((attr->srq == NULL) || (attr->srq_num != 0));
-    UCT_IB_MLX5DV_SET(qpc, qpc, rq_type, !!attr->srq_num);
-    UCT_IB_MLX5DV_SET(qpc, qpc, srqn_rmpn_xrqn, attr->srq_num);
+    ucs_assert((attr->super.srq == NULL) || (attr->super.srq_num != 0));
+    UCT_IB_MLX5DV_SET(qpc, qpc, rq_type, !!attr->super.srq_num);
+    UCT_IB_MLX5DV_SET(qpc, qpc, srqn_rmpn_xrqn, attr->super.srq_num);
     UCT_IB_MLX5DV_SET(qpc, qpc, cqn_snd, dvscq.cqn);
     UCT_IB_MLX5DV_SET(qpc, qpc, cqn_rcv, dvrcq.cqn);
     UCT_IB_MLX5DV_SET(qpc, qpc, log_sq_size, ucs_ilog2_or0(max_tx));
     UCT_IB_MLX5DV_SET(qpc, qpc, log_rq_size, ucs_ilog2_or0(max_rx));
     UCT_IB_MLX5DV_SET(qpc, qpc, cs_req, UCT_IB_MLX5_QPC_CS_REQ_UP_TO_64B);
     UCT_IB_MLX5DV_SET(qpc, qpc, cs_res,
-                      uct_ib_mlx5_qpc_cs_res(attr->max_inl_resp));
+                      uct_ib_mlx5_qpc_cs_res(attr->super.max_inl_resp));
     UCT_IB_MLX5DV_SET64(qpc, qpc, dbr_addr, qp->devx.dbrec->offset);
     UCT_IB_MLX5DV_SET(qpc, qpc, dbr_umem_id, qp->devx.dbrec->mem_id);
 
@@ -157,7 +157,7 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
     UCT_IB_MLX5DV_SET(rst2init_qp_in, in_2init, opcode, UCT_IB_MLX5_CMD_OP_RST2INIT_QP);
     UCT_IB_MLX5DV_SET(rst2init_qp_in, in_2init, qpn, qp->qp_num);
     UCT_IB_MLX5DV_SET(qpc, qpc, pm_state, UCT_IB_MLX5_QPC_PM_STATE_MIGRATED);
-    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.vhca_port_num, attr->port);
+    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.vhca_port_num, attr->super.port);
     UCT_IB_MLX5DV_SET(qpc, qpc, rwe, true);
 
     ret = mlx5dv_devx_obj_modify(qp->devx.obj, in_2init, sizeof(in_2init),
@@ -170,8 +170,8 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
 
     qp->type = UCT_IB_MLX5_OBJ_TYPE_DEVX;
 
-    attr->cap.max_send_wr = max_tx;
-    attr->cap.max_recv_wr = max_rx;
+    attr->super.cap.max_send_wr = max_tx;
+    attr->super.cap.max_recv_wr = max_rx;
 
     if (tx != NULL) {
         tx->reg    = &uar->super;

--- a/src/uct/ib/mlx5/dv/ib_mlx5_dv.c
+++ b/src/uct/ib/mlx5/dv/ib_mlx5_dv.c
@@ -50,6 +50,7 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
     char out[UCT_IB_MLX5DV_ST_SZ_BYTES(create_qp_out)]         = {};
     char in_2init[UCT_IB_MLX5DV_ST_SZ_BYTES(rst2init_qp_in)]   = {};
     char out_2init[UCT_IB_MLX5DV_ST_SZ_BYTES(rst2init_qp_out)] = {};
+    uct_ib_mlx5_mmio_mode_t mmio_mode;
     uct_ib_mlx5_devx_uar_t *uar;
     int max_tx, max_rx, len_tx, len;
     int wqe_size;
@@ -59,12 +60,18 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
 
     uct_ib_iface_fill_attr(iface, &attr->super);
 
+    status = uct_ib_mlx5_get_mmio_mode(iface->super.worker, attr->mmio_mode,
+                                       UCT_IB_MLX5_BF_REG_SIZE, &mmio_mode);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
     uar = uct_worker_tl_data_get(iface->super.worker,
                                  UCT_IB_MLX5_DEVX_UAR_KEY,
                                  uct_ib_mlx5_devx_uar_t,
                                  uct_ib_mlx5_devx_uar_cmp,
                                  uct_ib_mlx5_devx_uar_init,
-                                 md, UCT_IB_MLX5_MMIO_MODE_BF_POST);
+                                 md, mmio_mode);
     if (UCS_PTR_IS_ERR(uar)) {
         status = UCS_PTR_STATUS(uar);
         goto err;

--- a/src/uct/ib/mlx5/ib_mlx5.c
+++ b/src/uct/ib/mlx5/ib_mlx5.c
@@ -277,7 +277,7 @@ void uct_ib_mlx5_iface_put_res_domain(uct_ib_mlx5_qp_t *qp)
 
 ucs_status_t uct_ib_mlx5_iface_create_qp(uct_ib_iface_t *iface,
                                          uct_ib_mlx5_qp_t *qp,
-                                         uct_ib_qp_attr_t *attr)
+                                         uct_ib_mlx5_qp_attr_t *attr)
 {
     ucs_status_t status;
 
@@ -286,8 +286,8 @@ ucs_status_t uct_ib_mlx5_iface_create_qp(uct_ib_iface_t *iface,
         return status;
     }
 
-    uct_ib_exp_qp_fill_attr(iface, attr);
-    status = uct_ib_iface_create_qp(iface, attr, &qp->verbs.qp);
+    uct_ib_exp_qp_fill_attr(iface, &attr->super);
+    status = uct_ib_iface_create_qp(iface, &attr->super, &qp->verbs.qp);
     if (status != UCS_OK) {
         return status;
     }

--- a/src/uct/ib/mlx5/ib_mlx5.c
+++ b/src/uct/ib/mlx5/ib_mlx5.c
@@ -413,29 +413,6 @@ void uct_ib_mlx5_devx_uar_cleanup(uct_ib_mlx5_devx_uar_t *uar)
 #endif
 }
 
-ucs_status_t uct_ib_mlx5_txwq_init_devx(uct_priv_worker_t *worker,
-                                        uct_ib_mlx5_md_t *md,
-                                        uct_ib_mlx5_txwq_t *txwq,
-                                        uct_ib_mlx5_mmio_mode_t mode)
-{
-    uct_ib_mlx5_devx_uar_t *uar;
-
-    uar = uct_worker_tl_data_get(worker,
-                                 UCT_IB_MLX5_DEVX_UAR_KEY,
-                                 uct_ib_mlx5_devx_uar_t,
-                                 uct_ib_mlx5_devx_uar_cmp,
-                                 uct_ib_mlx5_devx_uar_init,
-                                 md, mode);
-    if (UCS_PTR_IS_ERR(uar)) {
-        return UCS_PTR_STATUS(uar);
-    }
-
-    txwq->reg        = &uar->super;
-    txwq->super.type = UCT_IB_MLX5_OBJ_TYPE_DEVX;
-
-    return UCS_OK;
-}
-
 void uct_ib_mlx5_txwq_reset(uct_ib_mlx5_txwq_t *txwq)
 {
     txwq->curr       = txwq->qstart;

--- a/src/uct/ib/mlx5/ib_mlx5.c
+++ b/src/uct/ib/mlx5/ib_mlx5.c
@@ -448,6 +448,32 @@ void uct_ib_mlx5_txwq_reset(uct_ib_mlx5_txwq_t *txwq)
     memset(txwq->qstart, 0, UCS_PTR_BYTE_DIFF(txwq->qstart, txwq->qend));
 }
 
+ucs_status_t
+uct_ib_mlx5_get_mmio_mode(uct_priv_worker_t *worker,
+                          uct_ib_mlx5_mmio_mode_t cfg_mmio_mode,
+                          unsigned bf_size,
+                          uct_ib_mlx5_mmio_mode_t *mmio_mode)
+{
+    ucs_assert(cfg_mmio_mode < UCT_IB_MLX5_MMIO_MODE_LAST);
+
+    if (cfg_mmio_mode != UCT_IB_MLX5_MMIO_MODE_AUTO) {
+        *mmio_mode = cfg_mmio_mode;
+    } else if (bf_size > 0) {
+        if (worker->thread_mode == UCS_THREAD_MODE_SINGLE) {
+            *mmio_mode = UCT_IB_MLX5_MMIO_MODE_BF_POST;
+        } else if (worker->thread_mode == UCS_THREAD_MODE_SERIALIZED) {
+            *mmio_mode = UCT_IB_MLX5_MMIO_MODE_BF_POST_MT;
+        } else {
+            ucs_error("unsupported thread mode for mlx5: %d", worker->thread_mode);
+            return UCS_ERR_UNSUPPORTED;
+        }
+    } else {
+        *mmio_mode = UCT_IB_MLX5_MMIO_MODE_DB;
+    }
+
+    return UCS_OK;
+}
+
 ucs_status_t uct_ib_mlx5_txwq_init(uct_priv_worker_t *worker,
                                    uct_ib_mlx5_mmio_mode_t cfg_mmio_mode,
                                    uct_ib_mlx5_txwq_t *txwq,
@@ -476,19 +502,10 @@ ucs_status_t uct_ib_mlx5_txwq_init(uct_priv_worker_t *worker,
         return UCS_ERR_IO_ERROR;
     }
 
-    if (cfg_mmio_mode != UCT_IB_MLX5_MMIO_MODE_AUTO) {
-        mmio_mode = cfg_mmio_mode;
-    } else if (qp_info.dv.bf.size > 0) {
-        if (worker->thread_mode == UCS_THREAD_MODE_SINGLE) {
-            mmio_mode = UCT_IB_MLX5_MMIO_MODE_BF_POST;
-        } else if (worker->thread_mode == UCS_THREAD_MODE_SERIALIZED) {
-            mmio_mode = UCT_IB_MLX5_MMIO_MODE_BF_POST_MT;
-        } else {
-            ucs_error("unsupported thread mode for mlx5: %d", worker->thread_mode);
-            return UCS_ERR_UNSUPPORTED;
-        }
-    } else {
-        mmio_mode = UCT_IB_MLX5_MMIO_MODE_DB;
+    status = uct_ib_mlx5_get_mmio_mode(worker, cfg_mmio_mode,
+                                       qp_info.dv.bf.size, &mmio_mode);
+    if (status != UCS_OK) {
+        return status;
     }
 
     ucs_debug("tx wq %d bytes [bb=%d, nwqe=%d] mmio_mode %s",

--- a/src/uct/ib/mlx5/ib_mlx5.h
+++ b/src/uct/ib/mlx5/ib_mlx5.h
@@ -516,11 +516,6 @@ ucs_status_t uct_ib_mlx5_txwq_init(uct_priv_worker_t *worker,
                                    uct_ib_mlx5_mmio_mode_t cfg_mmio_mode,
                                    uct_ib_mlx5_txwq_t *txwq, struct ibv_qp *verbs_qp);
 
-ucs_status_t uct_ib_mlx5_txwq_init_devx(uct_priv_worker_t *worker,
-                                        uct_ib_mlx5_md_t *md,
-                                        uct_ib_mlx5_txwq_t *txwq,
-                                        uct_ib_mlx5_mmio_mode_t mode);
-
 void uct_ib_mlx5_txwq_cleanup(uct_ib_mlx5_txwq_t* txwq);
 
 /**

--- a/src/uct/ib/mlx5/ib_mlx5.h
+++ b/src/uct/ib/mlx5/ib_mlx5.h
@@ -295,6 +295,7 @@ typedef struct uct_ib_mlx5_res_domain {
 
 typedef struct uct_ib_mlx5_qp_attr {
     uct_ib_qp_attr_t            super;
+    uct_ib_mlx5_mmio_mode_t     mmio_mode;
 } uct_ib_mlx5_qp_attr_t;
 
 
@@ -322,7 +323,6 @@ typedef struct uct_ib_mlx5_qp {
 #endif
     };
 } uct_ib_mlx5_qp_t;
-
 
 /* Send work-queue */
 typedef struct uct_ib_mlx5_txwq {
@@ -502,6 +502,12 @@ ucs_status_t uct_ib_mlx5dv_arm_cq(uct_ib_mlx5_cq_t *cq, int solicited);
  */
 void uct_ib_mlx5_check_completion(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq,
                                   struct mlx5_cqe64 *cqe);
+
+ucs_status_t
+uct_ib_mlx5_get_mmio_mode(uct_priv_worker_t *worker,
+                          uct_ib_mlx5_mmio_mode_t cfg_mmio_mode,
+                          unsigned bf_size,
+                          uct_ib_mlx5_mmio_mode_t *mmio_mode);
 
 /**
  * Initialize txwq structure.

--- a/src/uct/ib/mlx5/ib_mlx5.h
+++ b/src/uct/ib/mlx5/ib_mlx5.h
@@ -293,6 +293,11 @@ typedef struct uct_ib_mlx5_res_domain {
 } uct_ib_mlx5_res_domain_t;
 
 
+typedef struct uct_ib_mlx5_qp_attr {
+    uct_ib_qp_attr_t            super;
+} uct_ib_mlx5_qp_attr_t;
+
+
 /* MLX5 QP wrapper */
 typedef struct uct_ib_mlx5_qp {
     uct_ib_mlx5_obj_type_t             type;
@@ -465,7 +470,7 @@ void uct_ib_mlx5_iface_put_res_domain(uct_ib_mlx5_qp_t *qp);
 
 ucs_status_t uct_ib_mlx5_iface_create_qp(uct_ib_iface_t *iface,
                                          uct_ib_mlx5_qp_t *qp,
-                                         uct_ib_qp_attr_t *attr);
+                                         uct_ib_mlx5_qp_attr_t *attr);
 
 /**
  * Create CQ with DV
@@ -556,7 +561,7 @@ void uct_ib_mlx5_devx_uar_cleanup(uct_ib_mlx5_devx_uar_t *uar);
 ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
                                         uct_ib_mlx5_qp_t *qp,
                                         uct_ib_mlx5_txwq_t *tx,
-                                        uct_ib_qp_attr_t *attr);
+                                        uct_ib_mlx5_qp_attr_t *attr);
 
 ucs_status_t uct_ib_mlx5_devx_modify_qp(uct_ib_mlx5_qp_t *qp,
                                         const void *in, size_t inlen,
@@ -573,7 +578,7 @@ static inline ucs_status_t
 uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
                            uct_ib_mlx5_qp_t *qp,
                            uct_ib_mlx5_txwq_t *tx,
-                           uct_ib_qp_attr_t *attr)
+                           uct_ib_mlx5_qp_attr_t *attr)
 {
     return UCS_ERR_UNSUPPORTED;
 }

--- a/src/uct/ib/mlx5/ib_mlx5.inl
+++ b/src/uct/ib/mlx5/ib_mlx5.inl
@@ -509,7 +509,7 @@ uct_ib_mlx5_srq_get_wqe(uct_ib_mlx5_srq_t *srq, uint16_t wqe_index)
 static ucs_status_t UCS_F_MAYBE_UNUSED
 uct_ib_mlx5_iface_fill_attr(uct_ib_iface_t *iface,
                             uct_ib_mlx5_qp_t *qp,
-                            uct_ib_qp_attr_t *attr)
+                            uct_ib_mlx5_qp_attr_t *attr)
 {
     ucs_status_t status;
 
@@ -519,20 +519,20 @@ uct_ib_mlx5_iface_fill_attr(uct_ib_iface_t *iface,
     }
 
 #if HAVE_DECL_IBV_EXP_CREATE_QP
-    attr->ibv.comp_mask       = IBV_EXP_QP_INIT_ATTR_PD;
-    attr->ibv.pd              = uct_ib_iface_md(iface)->pd;
+    attr->super.ibv.comp_mask       = IBV_EXP_QP_INIT_ATTR_PD;
+    attr->super.ibv.pd              = uct_ib_iface_md(iface)->pd;
 #elif HAVE_DECL_IBV_CREATE_QP_EX
-    attr->ibv.comp_mask       = IBV_QP_INIT_ATTR_PD;
+    attr->super.ibv.comp_mask       = IBV_QP_INIT_ATTR_PD;
     if (qp->verbs.rd->pd != NULL) {
-        attr->ibv.pd          = qp->verbs.rd->pd;
+        attr->super.ibv.pd          = qp->verbs.rd->pd;
     } else {
-        attr->ibv.pd          = uct_ib_iface_md(iface)->pd;
+        attr->super.ibv.pd          = uct_ib_iface_md(iface)->pd;
     }
 #endif
 
 #ifdef HAVE_IBV_EXP_RES_DOMAIN
-    attr->ibv.comp_mask      |= IBV_EXP_QP_INIT_ATTR_RES_DOMAIN;
-    attr->ibv.res_domain      = qp->verbs.rd->ibv_domain;
+    attr->super.ibv.comp_mask      |= IBV_EXP_QP_INIT_ATTR_RES_DOMAIN;
+    attr->super.ibv.res_domain      = qp->verbs.rd->ibv_domain;
 #endif
 
     return UCS_OK;

--- a/src/uct/ib/rc/accel/rc_mlx5.h
+++ b/src/uct/ib/rc/accel/rc_mlx5.h
@@ -118,7 +118,7 @@ ucs_status_t uct_rc_mlx5_ep_fc_ctrl(uct_ep_t *tl_ep, unsigned op,
 ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
                                          uct_ib_mlx5_qp_t *qp,
                                          uct_ib_mlx5_txwq_t *txwq,
-                                         uct_ib_qp_attr_t *attr);
+                                         uct_ib_mlx5_qp_attr_t *attr);
 
 ucs_status_t uct_rc_mlx5_ep_connect_to_ep(uct_ep_h tl_ep,
                                           const uct_device_addr_t *dev_addr,

--- a/src/uct/ib/rc/accel/rc_mlx5_common.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_common.c
@@ -209,6 +209,7 @@ uct_rc_mlx5_devx_create_cmd_qp(uct_rc_mlx5_iface_common_t *iface)
     attr.super.ibv.recv_cq      = iface->super.super.cq[UCT_IB_DIR_RX];
     attr.super.srq_num          = iface->rx.srq.srq_num;
     attr.super.port             = dev->first_port;
+    attr.mmio_mode              = iface->tx.mmio_mode;
     status = uct_ib_mlx5_devx_create_qp(&iface->super.super,
                                         &iface->tm.cmd_wq.super.super,
                                         &iface->tm.cmd_wq.super,

--- a/src/uct/ib/rc/accel/rc_mlx5_common.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_common.c
@@ -197,18 +197,18 @@ uct_rc_mlx5_devx_create_cmd_qp(uct_rc_mlx5_iface_common_t *iface)
                                            uct_ib_mlx5_md_t);
     uct_ib_device_t *dev  = &md->super.dev;
     struct ibv_ah_attr ah_attr = {};
-    uct_ib_qp_attr_t attr = {};
+    uct_ib_mlx5_qp_attr_t attr = {};
     ucs_status_t status;
 
     ucs_assert(iface->tm.cmd_wq.super.super.type == UCT_IB_MLX5_OBJ_TYPE_LAST);
 
-    attr.cap.max_send_wr  = iface->tm.cmd_qp_len;
-    attr.cap.max_send_sge = 1;
-    attr.ibv.pd           = md->super.pd;
-    attr.ibv.send_cq      = iface->super.super.cq[UCT_IB_DIR_RX];
-    attr.ibv.recv_cq      = iface->super.super.cq[UCT_IB_DIR_RX];
-    attr.srq_num          = iface->rx.srq.srq_num;
-    attr.port             = dev->first_port;
+    attr.super.cap.max_send_wr  = iface->tm.cmd_qp_len;
+    attr.super.cap.max_send_sge = 1;
+    attr.super.ibv.pd           = md->super.pd;
+    attr.super.ibv.send_cq      = iface->super.super.cq[UCT_IB_DIR_RX];
+    attr.super.ibv.recv_cq      = iface->super.super.cq[UCT_IB_DIR_RX];
+    attr.super.srq_num          = iface->rx.srq.srq_num;
+    attr.super.port             = dev->first_port;
     status = uct_ib_mlx5_devx_create_qp(&iface->super.super,
                                         &iface->tm.cmd_wq.super.super,
                                         &iface->tm.cmd_wq.super,
@@ -219,7 +219,7 @@ uct_rc_mlx5_devx_create_cmd_qp(uct_rc_mlx5_iface_common_t *iface)
 
     ah_attr.is_global     = 1;
     ah_attr.grh.dgid      = iface->super.super.gid_info.gid;
-    ah_attr.dlid          = uct_ib_device_port_attr(dev, attr.port)->lid;
+    ah_attr.dlid          = uct_ib_device_port_attr(dev, attr.super.port)->lid;
     ah_attr.port_num      = dev->first_port;
     status = uct_rc_mlx5_iface_common_devx_connect_qp(
             iface, &iface->tm.cmd_wq.super.super,
@@ -443,23 +443,23 @@ void uct_rc_mlx5_iface_common_tag_cleanup(uct_rc_mlx5_iface_common_t *iface)
 }
 
 void uct_rc_mlx5_iface_fill_attr(uct_rc_mlx5_iface_common_t *iface,
-                                 uct_ib_qp_attr_t *qp_attr,
+                                 uct_ib_mlx5_qp_attr_t *qp_attr,
                                  unsigned max_send_wr,
                                  uct_ib_mlx5_srq_t *srq)
 {
     switch (srq->type) {
     case UCT_IB_MLX5_OBJ_TYPE_VERBS:
-        uct_rc_iface_fill_attr(&iface->super, qp_attr, max_send_wr,
+        uct_rc_iface_fill_attr(&iface->super, &qp_attr->super, max_send_wr,
                                srq->verbs.srq);
         break;
     case UCT_IB_MLX5_OBJ_TYPE_DEVX:
-        uct_rc_iface_fill_attr(&iface->super, qp_attr, max_send_wr, NULL);
+        uct_rc_iface_fill_attr(&iface->super, &qp_attr->super, max_send_wr, NULL);
         break;
     case UCT_IB_MLX5_OBJ_TYPE_LAST:
         break;
     }
 
-    qp_attr->srq_num = srq->srq_num;
+    qp_attr->super.srq_num = srq->srq_num;
 }
 
 static ucs_status_t

--- a/src/uct/ib/rc/accel/rc_mlx5_common.h
+++ b/src/uct/ib/rc/accel/rc_mlx5_common.h
@@ -722,7 +722,7 @@ uct_rc_mlx5_iface_common_devx_connect_qp(uct_rc_mlx5_iface_common_t *iface,
 #endif
 
 void uct_rc_mlx5_iface_fill_attr(uct_rc_mlx5_iface_common_t *iface,
-                                 uct_ib_qp_attr_t *qp_attr,
+                                 uct_ib_mlx5_qp_attr_t *qp_attr,
                                  unsigned max_send_wr,
                                  uct_ib_mlx5_srq_t *srq);
 

--- a/src/uct/ib/rc/accel/rc_mlx5_ep.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_ep.c
@@ -867,13 +867,13 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
 {
     uct_rc_mlx5_iface_common_t *iface = ucs_derived_of(params->iface,
                                                        uct_rc_mlx5_iface_common_t);
-    uct_ib_qp_attr_t attr = {};
+    uct_ib_mlx5_qp_attr_t attr = {};
     ucs_status_t status;
 
     /* Need to create QP before super constructor to get QP number */
     uct_rc_mlx5_iface_fill_attr(iface, &attr, iface->super.config.tx_qp_len,
                                 &iface->rx.srq);
-    uct_ib_exp_qp_fill_attr(&iface->super.super, &attr);
+    uct_ib_exp_qp_fill_attr(&iface->super.super, &attr.super);
     status = uct_rc_mlx5_iface_create_qp(iface, &self->tx.wq.super, &self->tx.wq, &attr);
     if (status != UCS_OK) {
         return status;
@@ -896,7 +896,7 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
          * such a QP to be initialized with zero send queue length. */
         memset(&attr, 0, sizeof(attr));
         uct_rc_mlx5_iface_fill_attr(iface, &attr, 0, &iface->rx.srq);
-        uct_ib_exp_qp_fill_attr(&iface->super.super, &attr);
+        uct_ib_exp_qp_fill_attr(&iface->super.super, &attr.super);
         status = uct_rc_mlx5_iface_create_qp(iface, &self->tm_qp, NULL, &attr);
         if (status != UCS_OK) {
             goto err;

--- a/src/uct/ib/rc/accel/rc_mlx5_iface.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_iface.c
@@ -250,7 +250,7 @@ static void uct_rc_mlx5_iface_progress_enable(uct_iface_h tl_iface, unsigned fla
 ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
                                          uct_ib_mlx5_qp_t *qp,
                                          uct_ib_mlx5_txwq_t *txwq,
-                                         uct_ib_qp_attr_t *attr)
+                                         uct_ib_mlx5_qp_attr_t *attr)
 {
     uct_ib_iface_t *ib_iface           = &iface->super.super;
     ucs_status_t status;
@@ -269,12 +269,12 @@ ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
         return status;
     }
 
-    uct_ib_iface_fill_attr(ib_iface, attr);
+    uct_ib_iface_fill_attr(ib_iface, &attr->super);
 #if HAVE_DECL_MLX5DV_QP_CREATE_ALLOW_SCATTER_TO_CQE
     dv_attr.comp_mask    = MLX5DV_QP_INIT_ATTR_MASK_QP_CREATE_FLAGS;
     dv_attr.create_flags = MLX5DV_QP_CREATE_ALLOW_SCATTER_TO_CQE;
 #endif
-    qp->verbs.qp = mlx5dv_create_qp(dev->ibv_context, &attr->ibv, &dv_attr);
+    qp->verbs.qp = mlx5dv_create_qp(dev->ibv_context, &attr->super.ibv, &dv_attr);
     if (qp->verbs.qp == NULL) {
         ucs_error("mlx5dv_create_qp("UCT_IB_IFACE_FMT"): failed: %m",
                   UCT_IB_IFACE_ARG(ib_iface));
@@ -295,7 +295,7 @@ ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
         goto err_destory_qp;
     }
 
-    if (attr->cap.max_send_wr) {
+    if (attr->super.cap.max_send_wr) {
         status = uct_ib_mlx5_txwq_init(iface->super.super.super.worker,
                                        iface->tx.mmio_mode, txwq,
                                        qp->verbs.qp);

--- a/src/uct/ib/rc/accel/rc_mlx5_iface.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_iface.c
@@ -261,6 +261,7 @@ ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
     struct mlx5dv_qp_init_attr dv_attr = {};
 
     if (md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_RC_QP) {
+        attr->mmio_mode = iface->tx.mmio_mode;
         return uct_ib_mlx5_devx_create_qp(ib_iface, qp, txwq, attr);
     }
 

--- a/src/uct/ib/ud/accel/ud_mlx5.c
+++ b/src/uct/ib/ud/accel/ud_mlx5.c
@@ -709,14 +709,17 @@ static void uct_ud_mlx5_iface_event_cq(uct_ib_iface_t *ib_iface,
 }
 
 static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
-                                                uct_ib_qp_attr_t *attr,
+                                                uct_ib_qp_attr_t *ib_attr,
                                                 struct ibv_qp **qp_p)
 {
     uct_ud_mlx5_iface_t *iface = ucs_derived_of(ib_iface, uct_ud_mlx5_iface_t);
     uct_ib_mlx5_qp_t *qp = &iface->tx.wq.super;
+    uct_ib_mlx5_qp_attr_t attr = {};
     ucs_status_t status;
 
-    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, attr);
+    attr.super     = *ib_attr;
+
+    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, &attr);
     if (status != UCS_OK) {
         return status;
     }

--- a/src/uct/ib/ud/accel/ud_mlx5.c
+++ b/src/uct/ib/ud/accel/ud_mlx5.c
@@ -718,6 +718,7 @@ static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
     ucs_status_t status;
 
     attr.super     = *ib_attr;
+    attr.mmio_mode = UCT_IB_MLX5_MMIO_MODE_LAST;
 
     status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, &attr);
     if (status != UCS_OK) {

--- a/test/gtest/uct/ib/test_devx.cc
+++ b/test/gtest/uct/ib/test_devx.cc
@@ -48,16 +48,4 @@ UCS_TEST_P(test_devx, dbrec)
     ucs_mpool_put_inline(dbrec);
 }
 
-UCS_TEST_P(test_devx, uar)
-{
-    uct_ib_mlx5_txwq_t txwq;
-    ucs_status_t status;
-
-    status = uct_ib_mlx5_txwq_init_devx(worker(), md(), &txwq,
-                                        UCT_IB_MLX5_MMIO_MODE_BF_POST);
-    ASSERT_UCS_OK(status);
-    ASSERT_TRUE(txwq.super.type == UCT_IB_MLX5_OBJ_TYPE_DEVX);
-    uct_ib_mlx5_txwq_cleanup(&txwq);
-}
-
 UCT_INSTANTIATE_IB_TEST_CASE(test_devx);

